### PR TITLE
Adding ViewModel Tests

### PR DESCRIPTION
**Pull request by Dearop**
# PR Title: Adding ViewModel Unit Tests and Enhancements

## Summary 

This PR introduces a comprehensive suite of unit tests for various ViewModel classes, improving the test coverage and robustness of the application. These tests focus on ensuring the correctness of ViewModel behaviors and interactions with repositories. Additionally, minor improvements were made to ViewModel implementations to align with best practices.

### 1. Added Unit Tests:
* Comprehensive tests for BarcodeScannerViewModel, InvitationViewModel, SignInViewModel, OverviewScreenViewModel, and more.
* Utilized Hilt for dependency injection in test cases to provide isolated and mock dependencies.
* Tested scenarios such as error handling, state changes, and repository interactions.
### 2. Code Adjustments:
* Updated certain suspend functions to non-suspend where appropriate (e.g., deleteInvitationUID) for consistency.
*Enhanced error handling to prevent app crashes during repository updates (e.g., updateStinkyPoints).
* Modified lifecycle usage in BarcodeScannerViewModel to better align with lifecycle-aware components.
### 3. New Test Modules:
* Added test modules for CoroutineScope, FoodFactsRepository, UserRepository, and more to enable dependency injection in test environments.
* Replaced production modules in tests using @TestInstallIn.
### 4.	Gradle Updates:
* Added dependencies for androidx.lifecycle.runtime.testing and androidx.core.testing.
* Updated versions for various test libraries to ensure compatibility and stability.

### Production Code:
* ViewModel classes: Adjustments to signatures and lifecycle-aware patterns.
* Repositories: Minor bug fixes and error handling enhancements.
### Test Code:
* test/java/... - Added extensive unit tests for ViewModels and Repositories.
* New test modules for dependency injection.
### Gradle:
* build.gradle.kts: Added new dependencies for testing.
* libs.versions.toml: Updated versions and added new library definitions.

### BarcodeScannerViewModelTest:
* Validated camera permission handling and repository interactions.
* Verified correct state updates during barcode scanning.
### InvitationViewModelTest:
* Ensured proper handling of accepting and declining invitations.
* Validated error scenarios in refreshInvitations.
### OverviewScreenViewModelTest:
* Verified correct filter management and food item selection.
* Simulated food item status updates and household interactions.
### SignInViewModelTest:
* Tested successful and failed sign-in scenarios.
* Verified error states and state transitions.

## Additional Notes

* This PR lays the foundation for further testing improvements, including UI testing.
* The introduced tests should significantly reduce regression risks during future updates.

## Next Steps

* Review and merge this PR to integrate improved test coverage.
* Continue adding tests for remaining classes and edge cases.

### CODE REVIEW

File: app/src/main/java/com/android/shelfLife/viewmodel/camera/BarcodeScannerViewModel.kt [hunk 1]
```diff
@@ -26,17 +26,16 @@ import kotlinx.coroutines.flow.StateFlow
...
+) : ViewModel(), LifecycleObserver {
 
-  private val application = context as Application
   private val sharedPreferences =
       application.getSharedPreferences("app_prefs", Context.MODE_PRIVATE)
 
   var permissionGranted by
       mutableStateOf(
-          ContextCompat.checkSelfPermission(context, Manifest.permission.CAMERA) ==
+          ContextCompat.checkSelfPermission(application, Manifest.permission.CAMERA) ==
```
Comment (frostymate) [inline app/src/main/java/com/android/shelfLife/viewmodel/camera/BarcodeScannerViewModel.kt:38]: good use of refactoring variables that do not need to be renamed

File: app/src/main/java/com/android/shelfLife/viewmodel/invitations/InvitationViewModel.kt [hunk 2]
```diff
@@ -46,13 +46,14 @@ constructor(
...
 
-  private suspend fun refreshInvitations() {
+  internal suspend fun refreshInvitations() {
     val invitationUIDs = userRepo.invitations.value
     if (invitationUIDs.isNotEmpty()) {
       try {
         _invitations.value = invitationRepository.getInvitationsBatch(invitationUIDs)
       } catch (e: Exception) {
         Log.e("InvitationViewModel", "Error while refreshing invitations", e)
+        _invitations.value = emptyList()
```
Comment (frostymate) [inline app/src/main/java/com/android/shelfLife/viewmodel/invitations/InvitationViewModel.kt:56]: Nice catching this error, now bugs or undefined behaviour will not be introduced if there is a exception

File: app/src/main/java/com/android/shelfLife/viewmodel/overview/IndividualFoodItemViewModel.kt [hunk 3]
```diff
@@ -25,7 +25,7 @@ constructor(
 
   suspend fun deleteFoodItem() {
     val householdId = userRepository.user.value?.selectedHouseholdUID
-    if (householdId != null) {
+    if (householdId != null && selectedFood != null) {
```
Comment (frostymate) [inline app/src/main/java/com/android/shelfLife/viewmodel/overview/IndividualFoodItemViewModel.kt:28]: This is a good catch as a food has to be selected and valid for it to be deleted

File: app/src/test/java/com/android/shelflife/di/TestRecipeModule.kt [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+import com.android.shelfLife.model.recipe.RecipeRepositoryFirestore
+import com.google.firebase.firestore.FirebaseFirestore
+import dagger.Module
+import dagger.Provides
+import dagger.hilt.components.SingletonComponent
+import dagger.hilt.testing.TestInstallIn
+import javax.inject.Singleton
+
+@Module
+@TestInstallIn(components = [SingletonComponent::class], replaces = [RecipeRepositoryModule::class])
```
Comment (frostymate) [inline app/src/test/java/com/android/shelflife/di/TestRecipeModule.kt:14]: Great job figuring how to mock the recipes and other repositories, it was something that the team was stuck on for a while

File: app/src/test/java/com/android/shelflife/model/HouseholdRepositoryFirestoreTest.kt [hunk 5]
```diff
@@ -245,4 +248,286 @@ class HouseholdRepositoryFirestoreTest {
...
+    assertTrue(houseHoldRepository.households.value.isEmpty())
+    assertNull(houseHoldRepository.selectedHousehold.value)
+  }
+
+  @Test
+  fun addHouseholdErrorRollsBackChanges() = runBlocking {
+    val household =
+        HouseHold("houseErr", "Error Household", listOf("u1"), listOf("r1"), emptyMap(), emptyMap())
+
+    `when`(mockCollection.document(household.uid)).thenReturn(mockDocument)
```
Comment (frostymate) [inline app/src/test/java/com/android/shelflife/model/HouseholdRepositoryFirestoreTest.kt:376]: are these when statements specific for this test or they can be put in the setup?

File: app/src/test/java/com/android/shelflife/model/HouseholdRepositoryFirestoreTest.kt [hunk 6]
```diff
@@ -245,4 +248,286 @@ class HouseholdRepositoryFirestoreTest {
...
+  }
+
+  @Test
+  fun `initializeHouseholds handles Firestore error gracefully`(): Unit = runBlocking {
+    // Arrange
+    val householdIds = listOf("house123", "house456")
+    val mockQuery = mock(Query::class.java)
+
+    `when`(mockCollection.whereIn(FieldPath.documentId(), householdIds)).thenReturn(mockQuery)
+    `when`(mockQuery.get()).thenThrow(RuntimeException("Test Initialization Error"))
```
Comment (frostymate) [inline app/src/test/java/com/android/shelflife/model/HouseholdRepositoryFirestoreTest.kt:521]: One small qualm I have is that these when statements are repeated or are very close in each test therefore in order to prevent repeated code and make the tests cleaner one could have them declared in the set up

**Review by frostymate — APPROVED** [review]
Paul Great Job on writing the tests especially for the view Model. I know it was tough as we had not done view model test previously by you managed to pull through and they seem to be passing the CI. I guess what I wanted to say that the viewModel tests more the functionality of the tests/logic while the UI/ Unit tests are more about the correct composables being displayed in a screen which is fine. One small improvement that I would suggest and for which I have written some comments for is that I think that for some tests some "when" statements have repeated functionality, I think the code could be improving by moving this logic to the set up where it could be avaliable for all test in the test file. Overall Great Job just make sure that for the View Model things are injected correctly and that they are able to be "mocked" correctly.


2. **app/src/main/java/com/android/shelfLife/model/household/HouseHoldRepositoryFirestore.kt**:
   - **Changes**:
     - Added a variable `oldStinkyPoints` to store previous stinky points before updating them.
   - **Review**:
     - The addition of `oldStinkyPoints` is a good practice for rollback or debugging purposes if needed.
     - Ensure `oldStinkyPoints` is used appropriately or cleaned up if not needed to avoid unnecessary memory usage.

3. **app/src/main/java/com/android/shelfLife/model/user/UserRepository.kt**:
   - **Changes**:
     - Changed `deleteInvitationUID` from a `suspend` function to a regular function.
   - **Review**:
     - This change might have implications on the calling code. Ensure that this method is not called from a coroutine context where suspension is required.

4. **app/src/main/java/com/android/shelfLife/model/user/UserRepositoryFirestore.kt**:
   - **Changes**:
     - Changed `updateArrayField` and `deleteInvitationUID` from `suspend` functions to regular functions.
   - **Review**:
     - Similar to the previous file, ensure that these changes do not affect the coroutine flow and are compatible with non-suspending contexts.

5. **app/src/main/java/com/android/shelfLife/viewmodel/camera/BarcodeScannerViewModel.kt**:
   - **Changes**:
     - Replaced `Context` with `Application` and added `LifecycleObserver`.
   - **Review**:
     - Using `Application` is often a better practice for long-lived ViewModels.
     - Adding `LifecycleObserver` improves lifecycle management, ensuring the ViewModel reacts to lifecycle events properly.

6. **app/src/main/java/com/android/shelfLife/viewmodel/invitations/InvitationViewModel.kt**:
   - **Changes**:
     - Changed `refreshInvitations` to be `internal` and added setting `_invitations` to an empty list in case of an error.
   - **Review**:
     - Making `refreshInvitations` internal restricts its visibility, which is good for encapsulation.
     - Handling errors by setting `_invitations` to an empty list ensures the ViewModel state remains consistent even on failure.

7. **app/src/main/java/com/android/shelfLife/viewmodel/overview/IndividualFoodItemViewModel.kt**:
   - **Changes**:
     - Added a check for `selectedFood` being non-null before deleting a food item.
   - **Review**:
     - This is a good null check to prevent potential `NullPointerException`.

8. **app/src/main/java/com/android/shelfLife/viewmodel/overview/OverviewScreenViewModel.kt**:
   - **Changes**:
     - Changed visibility of `checkItemStatus` to `private`.
   - **Review**:
     - Restricting the visibility of `checkItemStatus` to private is good practice to enforce encapsulation.

9. **app/src/main/java/com/android/shelfLife/viewmodel/profile/ProfileScreenViewModel.kt**:
   - **Changes**:
     - Removed `InvitationRepository` from the constructor.
   - **Review**:
     - If `InvitationRepository` is no longer used in this ViewModel, removing it is a good clean-up step.

10. **New Test Modules**:
    - Added new test modules for `TestCoroutineScopeModule`, `TestFoodFactsModule`, and `TestRecipeModule`.
    - **Review**:
      - Adding these test modules enhances the testability of the corresponding components, which is a positive step towards ensuring code quality.

### General Feedback
- The changes are generally well-structured and improve the codebase in terms of testing, encapsulation, and lifecycle management.
- Ensure that all changes, especially those changing from `suspend` to non-suspend functions, have been tested to verify that they do not introduce any unintended side effects.
- Adding corresponding unit tests for the changes will help in validating their correctness and robustness.

For further details, you can view the pull request files [here](https://github.com/commit-crimes/shelf-life/pull/320/files).
